### PR TITLE
Add CUDA aware support for UCX and OpenMPI

### PR DESCRIPTION
**Pull request by Harry-Chen**
According to https://www.open-mpi.org/faq/?category=buildcuda, OpenMPI could be made CUDA-aware by using UCX built with `gdrcopy` and `CUDA`.

Although ucx is able to detect the existence of CUDA and gdrcopy when configuring, I believe an explicit flag would be more informative.

BTW it there a way to express that when `cuda_aware` is enabled, `fabrics` must be `ucx`? The current implementation is ugly for I have to include all other fabrics in `conflicts` section explicitly.

### CODE REVIEW

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 1]
```diff
@@ -44,4 +51,9 @@ def configure_args(self):
             config_args.append('--enable-mt')
         else:
             config_args.append('--disable-mt')
+        if '+cuda' in spec:
+            config_args.append('--with-cuda={0}'.format(spec['cuda'].prefix))
+        if '+gdrcopy' in spec:
+            config_args.append('--with-gdrcopy')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ucx/package.py]: Can these be disabled with `--without`?

File: var/spack/repos/builtin/packages/ucx/package.py [hunk 2]
```diff
@@ -44,4 +51,9 @@ def configure_args(self):
...
         return config_args
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ucx/package.py]: This line will cause flake8 to fail

File: var/spack/repos/builtin/packages/ucx/package.py (hunk 1 above)
Comment (Harry-Chen) [reply var/spack/repos/builtin/packages/ucx/package.py]: There are actually no 'without' options for these flags, and specifying no 'with' options in effect disable them

**Comment by Harry-Chen** [comment]
Ahh, sorry for the confusion. I just realized that the cuda flag in OpenMPI refers to CUDA-awareness. The problem might just be that it does not pass this dependency to ucx (by enabling the two flags I added).

 It might be more appropriate not to add the cuda_aware flag in OpenMPI but to make 'fabrics=ucx +cuda' depends on 'ucx +cuda +gdrcopy'.

**Comment by Harry-Chen** [comment]
To make it more clear, since OpenMPI 2.0.0, CUDA support is provided by UCX instead of OpenMPI itself. That is, +cuda flag in OpenMPI requires either of:

1. version < 2.0.0
2. version >= 2.0.0 with fabrcis=ucx ^ucx +cuda +gdrcopy

I will try to rewrite the dependency rules to ensure these constraints.

**Comment by Harry-Chen** [comment]
I have now removed the `cuda_aware` flag, and corrent the dependency of `cuda` flag in the `openmpi` package. Could you please review now @hppritcha

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/openmpi/package.py:299: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:301: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:303: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:305: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:307: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:309: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/openmpi/package.py:311: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/ucx/package.py:41: [E501] line too long (88 > 79 characters)
```

**Comment by Harry-Chen** [comment]
> I suspect the gdr-copy variant is likely going to get some complaints from the container crews. You're sort of locking in those trying to build portable container stacks to only building on systems where gdr-copy device driver is installed. I''m not sure what to suggest there though. I guess they can make UCX a no build package and go with what's on their build systems.

**Comment by Harry-Chen** [comment]
>  For instance, the existing cuda variant is sufficient for PSM2 to be initialized to support whatever CUDA aware functionality there is inside PSM2. Likewise, at some point in the relatively near future, there will be better CUDA support in libfabric. Perhaps a variant name like cuda_ucx?

Thanks for your notice.

> You're sort of locking in those trying to build portable container stacks to only building on systems where gdr-copy device driver is installed. 

Actually not. I tried and `--with-gdrcopy` flag will not cause the configuration process to fail even those it does not exist. But I must admit that a `ucx +gdrcopy` with no `gdrcopy` found is really confusing.

**Comment by Harry-Chen** [comment]
I believe I should look for a more appropriate way for this. I'm just gonna leave this PR closed before I could think of any new way.

**Comment by hppritcha** [comment]
@Harry-Chen maybe open one PR against UCX for GPU optimization? Then those installing OMPI who want CUDA support for their Mellanox IB cluster could then specify to build the optimized ucx variant and have the open mpi spec for their build request that variant of UCX plus the existing cuda variant for Open MPI.

**Comment by hppritcha** [comment]
I'd be willing to give that a shot on one of our Nvidia/IB clusters.

**Comment by Harry-Chen** [comment]
> @Harry-Chen maybe open one PR against UCX for GPU optimization? Then those installing OMPI who want CUDA support for their Mellanox IB cluster could then specify to build the optimized ucx variant and have the open mpi spec for their build request that variant of UCX plus the existing cuda variant for Open MPI.

Sure, I will look into that.

**Pull request closed without merging**